Various consistency updates, some fixes, mostly minor (#1169)
ietf-tapswg/api-drafts: draft-ietf-taps-impl.md

informative:
    I-D.ietf-quic-transport:

>     I-D.ietf-tcpm-2140bis:

Author: This is now RFC 9040.

Reviewer: Can you also fix I-D.ietf-quic-transport to be RFC9000?

Author: ok!

- the Connection, the basic object that represents a flow of data as Messages in either direction between the Local and Remote Endpoints;
- and the Listener, a passive waiting object that delivers new Connections.

> Preconnection objects should be implemented as bundles of properties that an application can both read and write. A Preconnection object influences a Connection only at one point in time: when the Connection is created. Connection objects represent the interface between the application and the implementation to manage transport state, and conduct data transfer. During the process of establishment ({{conn-establish}}), the Connection will not necessarily be immediately bound to a specific transport protocol instance, since multiple candidate Protocol Stacks might be raced.

Author: I mean, it's legal to implement a Transport Services system that immediately binds to TCP and TCP only, I guess...

Reviewer 2: is /specific/ needed if we say this?

Author: ```suggestion
Preconnection objects should be implemented as bundles of properties that an application can both read and write. A Preconnection object influences a Connection only at one point in time: when the Connection is created. Connection objects represent the interface between the application and the implementation to manage transport state, and conduct data transfer. During the process of establishment ({{conn-establish}}), the Connection will not necessarily be immediately bound to a transport protocol instance, since multiple candidate Protocol Stacks might be raced.
```

# Implementing Pre-Establishment

> The pre-establishment phase allows applications to specify properties for the Connections that they are about to make, or to query the API about potential Connections they could make.

Reviewer: "Pre-Establishment" is capitalized in the architecture doc, btw. We may want to change for consistency.

> During pre-establishment the application specifies one or more Endpoints to be used for communication as well as protocol preferences and constraints via Selection Properties and, if desired, also Connection Properties. Section 4 of {{I-D.ietf-taps-interface}} states that Connection Properties should preferably be configured during pre-establishment, because they can serve as input to decisions that are made by the implementation (e.g., the capacity profile can guide usage of a protocol offering scavenger-type congestion control).

Author: This clarifies that the prescription is in the interface draft, not here, and it's a bit more precise than "as early as possible"

The implementation stores these properties as a part of the Preconnection object for use during connection establishment. For Selection Properties that are not provided by the application, the implementation uses the default values specified in the Transport Services API ({{I-D.ietf-taps-interface}}).

## Configuration-time errors

The Transport Services system should have a list of supported protocols available, which each have transport features reflecting the capabilities of the protocol. Once an application specifies its Transport Properties, the Transport Services system matches the required and prohibited properties against the transport features of the available protocols (see Section 6.2 of {{I-D.ietf-taps-interface}} for the definition of property preferences).

In the following cases, failure should be detected during pre-establishment:

- A request by an application for properties that cannot be satisfied by any of the available protocols. For example, if an application requires `perMsgReliability`, but no such feature is available in any protocol on the host running the Transport Services system this should result in an error.

> - A request by an application for properties that are in conflict with each other. Required and prohibited properties cannot be satisfied by the same protocol. For example, if an application prohibits `reliability` but then requires `perMsgReliability`, this mismatch should result in an error.

Reviewer: ```suggestion
- A request by an application for properties that are in conflict with each other, such as specifying required and prohibited properties that cannot be satisfied by any protocol. For example, if an application prohibits `reliability` but then requires `perMsgReliability`, this mismatch should result in an error.
```

### Implementing listeners for Multiplexed Protocols
### Implementing Listeners for Multiplexed Protocols

Protocols that provide multiplexing of streams into a single four-tuple can listen both for entirely new connections (a new HTTP/2 stream on a new TCP connection, for example) and for new sub-connections (a new HTTP/2 stream on an existing connection). If the abstraction of Connection presented to the application is mapped to the multiplexed stream, then for either case the Listener ought to deliver the new Connection objects in the same way. The implementation should allow the application to introspect the Connection Group marked on the Connections to determine the grouping of the multiplexing.

# Implementing Sending and Receiving Data

> The most basic mapping for sending a Message is an abstraction of datagrams, in which the transport protocol naturally deals in discrete packets (such as UDP). Each Message here corresponds to a single datagram. Generally, these will be short enough that sending and receiving will always use a complete Message.

Author: I don't understand the last sentence, and it seems unnecessary to me.

- `msgChecksumLen`: when this is set to any value other than `Full Coverage`, it sets the minimum protection in protocols that allow limiting the checksum length (e.g. UDP-Lite). If the Protocol Stack does not support checksum length limitation, this property may be ignored.

- `msgReliable`: When true, the property specifies that the Message must be reliably transmitted. When false, and if unreliable transmission is supported by the underlying protocol, then the Message should be unreliably transmitted. If the underlying
protocol does not support unreliable transmission, the Message should be reliably transmitted.

- `msgCapacityProfile`: When true, this expresses a wish to override the
Generic Connection Property `connCapacityProfile` for this Message. Depending on the
value, this can, for example, be implemented by changing the DSCP value of the
associated packet (note that the guidelines in Section 6 of {{?RFC7657}} apply; e.g.,
the DSCP value should not be changed for different packets within a reliable
transport protocol session or DCCP connection).

> - `noFragmentation`: Setting this avoids network-layer fragmentation. Messages exceeding the transport’s current estimate of its maximum packet size (the `singularTransmissionMsgMaxLen` Connection Property) can result in transport segmentation when permitted, or generate an error. When used with transports running over IP version 4, the Don't Fragment bit should be set to avoid on-path IP fragmentation ({{!RFC8304}}).

Author: This is a guide for developers, not a list of facts about how existing systems behave - that was my rationale here.

The set of Connection Properties that are supported for setting and getting on a Connection are described in {{I-D.ietf-taps-interface}}. For
any properties that are generic, and thus could apply to all protocols being used by a Connection, the Transport Services implementation should store the properties

> in storage common to all protocols, and notify all protocol instances in the Protocol Stack whenever the properties have been modified by the application. {{!RFC8303}} and {{!RFC8304}} offer guidance on how to do this for TCP, MPTCP, SCTP, UDP and UDP-Lite; see {#specific-protocol-considerations} for a description of a back-tracking method to find the relevant protocol primitives using these documents.

Author: It just seemed natural to me to add "and this is how to do it" here.

Each protocol has a notion of Connectedness. Possible values for Connectedness are:

- Connectionless. Connectionless protocols do not establish explicit state between endpoints, and do not perform a handshake during Connection establishment.

> - Connected. Connected (also called "connection-oriented") protocols establish state between endpoints, and perform a handshake during connection establishment. The handshake may be 0-RTT to send data or resume a session, but bidirectional traffic is required to confirm connectedness.

Author: I added the "connection-oriented" thing because that term is used instead in the other documents, in some places. If we absolutely only want "Connected", then we should change these other occurrences to be consistent.
